src/datadog/w3c_propagation.cpp

```diff
@@ -0,0 +1,396 @@
+#include "w3c_propagation.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <regex>
+#include <utility>
+
+#include "dict_reader.h"
+#include "hex.h"
+#include "parse_util.h"
+#include "tags.h"
+
+namespace datadog {
+namespace tracing {
+namespace {
+
+// Return a predicate that returns whether its `char` argument is any of the
+// following:
+//
+// - outside of the ASCII inclusive range `[lowest_ascii, highest_ascii]`
+// - equal to one of the `disallowed_characters`.
+//
+// `verboten` is used with `std::replace_if` to sanitize field values within the
+// tracestate header.
+auto verboten(int lowest_ascii, int highest_ascii,
+              StringView disallowed_characters) {
+  return [=, chars = disallowed_characters](char ch) {
+    return int(ch) < lowest_ascii || int(ch) > highest_ascii ||
+           std::find(chars.begin(), chars.end(), ch) != chars.end();
+  };
+}
+
+// Populate the specified `result` with data extracted from the "traceparent"
+// entry of the specified `headers`. Return `nullopt` on success. Return a value
+// for the `tags::internal::w3c_extraction_error` tag if an error occurs.
+Optional<std::string> extract_traceparent(ExtractedData& result,
+                                          const DictReader& headers) {
+  const auto maybe_traceparent = headers.lookup("traceparent");
+  if (!maybe_traceparent) {
+    return nullopt;
+  }
+
+  const auto traceparent = strip(*maybe_traceparent);
+
+  // Note that leading and trailing whitespace was already removed above.
+  // Note that match group 0 is the entire match.
+  static const auto& pattern =
+      "([0-9a-f]{2})"  // hex version number (match group 1)
+      "-"
+      "([0-9a-f]{16}([0-9a-f]{16}))"  // hex trace ID (match groups 2 and 3)
+      "-"
+      "([0-9a-f]{16})"  // hex parent span ID (match group 4)
+      "-"
+      "([0-9a-f]{2})"  // hex "trace-flags" (match group 5)
+      "(?:$|-.*)";     // either the end, or a hyphen preceding further fields
+  static const std::regex regex{pattern};
+
+  std::match_results<StringView::iterator> match;
+  if (!std::regex_match(traceparent.begin(), traceparent.end(), match, regex)) {
+    return "malformed_traceparent";
+  }
+
+  assert(match.ready());
```

Comment:
Do you want any of these asserts checked at runtime, or not so much?

Reply:
I want them checked at runtime in debug builds, that's why I use the `assert` macro.